wiki_articles/compilation_steps.md

```diff
@@ -0,0 +1,26 @@
+# How Is C/C++ Source Code Compiled?
+
+\:hash: The **preprocessor** expands macros and copy-pastes the content of included files
+at their point of inclusion. After being preprocessed, a source file no longer
+contains preprocessor directives. The result is called a _translation unit_ (TU).
+
+> :bulb: _The preprocessor merely does text replacement, it does not care about any aspect of C/C++._
+
+:ce: The **compiler** turns instructions into machine code and enforces C/C++
+syntax rules. In particular, anything that is used must be have been declared
+beforehand. The compiler only needs to know that something **exists**, not how
+it is _defined_.
+
+> :bulb: _Headers are a convenient way to expose function declarations to the compiler. Function definitions can reside in another, separately compiled TU, allowing finer control over the compilation process._
+
+As function definitions might be unavailable at this point, the compiler emits
+dummy machine code for function calls. A final step must be performed to fix those.
+
+:link: The **linker** takes all compiled TUs at once and patches the dummy calls
+emitted by the compiler. For each of those, the linker finds the compiled
+function and inserts an address to jump to when the call is performed.
+
+> :bulb: _Errors stating `undefined reference` or `unresolved external symbol` are linker errors, and they occur when the linker is unable to find a definition within the TUs it was given._
+
+The linker also packs all the compiled code into one of three file formats:
+an executable, a static library, or a dynamic link library.
```

Comment:
Minor nit: A static library on linux isn't really linked it's just an archive of .o files (the program used to make a static library is `ar` as in *ar*chive and that happens to be the same "ar" in "tar"). I think this is the same on windows but I'd need to check.

Of course this is too in-the-weeds for a beginner article, but maybe wording it as `The linker packs the compiled code into the final executable or library.` would be a good simplification.

Reply:
Good to know, I had no idea! Putting in your rewording.